util/tracecpu/profile_test.go

```diff
@@ -0,0 +1,253 @@
+	tracecpu.GlobalSQLStatsProfiler.SetCollector(collector)
+	reqs := []struct {
+		sql  string
+		plan string
+	}{
+		{"select * from t where a=?", "point-get"},
+		{"select * from t where a>?", "table-scan"},
+		{"insert into t values (?)", ""},
+	}
+	var wg sync.WaitGroup
+	for _, req := range reqs {
+		wg.Add(1)
+		go func(sql, plan string) {
+			defer wg.Done()
+			s.mockExecuteSQL(sql, plan)
+		}(req.sql, req.plan)
+	}
+
+	cnt := collector.getCollectCnt()
+	// test for StartCPUProfile.
+	buf := bytes.NewBuffer(nil)
+	err := tracecpu.StartCPUProfile(buf)
+	c.Assert(err, IsNil)
+	s.waitCollectCnt(collector, cnt+2)
+	err = tracecpu.StopCPUProfile()
+	c.Assert(err, IsNil)
+	_, err = profile.Parse(buf)
+	c.Assert(err, IsNil)
+
+	// test for collect SQL stats.
+	wg.Wait()
+	for _, req := range reqs {
+		stats := collector.getSQLStats(req.sql, req.plan)
+		c.Assert(stats, NotNil)
+		sql := collector.getSQL(stats.SQLDigest)
+		plan := collector.getPlan(stats.PlanDigest)
+		c.Assert(sql, Equals, req.sql)
+		c.Assert(plan, Equals, req.plan)
+	}
+}
+
+func (s *testSuite) TestIsEnabled(c *C) {
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsTrue)
+	s.setTopSQLEnable(false)
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsFalse)
+
+	s.setTopSQLEnable(true)
+	err := tracecpu.StartCPUProfile(bytes.NewBuffer(nil))
+	c.Assert(err, IsNil)
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsTrue)
+	s.setTopSQLEnable(false)
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsTrue)
+	err = tracecpu.StopCPUProfile()
+	c.Assert(err, IsNil)
+
+	s.setTopSQLEnable(false)
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsFalse)
+	s.setTopSQLEnable(true)
+	c.Assert(tracecpu.GlobalSQLStatsProfiler.IsEnabled(), IsTrue)
+}
+
+func (s *testSuite) waitCollectCnt(collector *mockCollector, cnt int64) {
+	timeout := time.After(time.Second * 5)
+	for {
+		// Wait for collector collect sql stats count >= expected count
+		if collector.getCollectCnt() >= cnt {
+			break
+		}
+		select {
+		case <-timeout:
+			break
+		default:
+			time.Sleep(time.Millisecond * 10)
+		}
+	}
+}
+
+func (s *testSuite) setTopSQLEnable(enabled bool) {
+	cfg := config.GetGlobalConfig()
+	newCfg := *cfg
+	newCfg.TopSQL.Enable = enabled
+	config.StoreGlobalConfig(&newCfg)
+}
+
+func (s *testSuite) mockExecuteSQL(sql, plan string) {
+	ctx := context.Background()
+	sqlDigest := genDigest(sql)
+	ctx = tracecpu.SetGoroutineLabelsWithSQL(ctx, sql, sqlDigest)
+	s.mockExecute(time.Millisecond * 100)
+	planDigest := genDigest(plan)
+	tracecpu.SetGoroutineLabelsWithSQLAndPlan(ctx, sqlDigest, planDigest, plan)
+	s.mockExecute(time.Millisecond * 300)
+}
+
+func genDigest(str string) string {
+	if str == "" {
+		return ""
+	}
+	hasher := sha256.New()
+	hasher.Write(hack.Slice(str))
+	return hex.EncodeToString(hasher.Sum(nil))
+}
+
+func (s *testSuite) mockExecute(d time.Duration) {
+	start := time.Now()
+	for {
+		for i := 0; i < 10e5; i++ {
```

Comment:
This might be optimized to `nop` by the compiler, I'm not sure.

Reply:
Ok, but it doesn't matter, since there is a loop that only return when timeout.